modules/services/spotifyd.nix

```diff
@@ -0,0 +1,184 @@
+{ config, lib, pkgs, ... }:
+
+let
+  cfg = config.services.spotifyd;
+
+  configFile = pkgs.writeText "spotifyd.conf" ''
+    [global]
+    username = ${cfg.user}
+    ${lib.optionalString (cfg.password != null) "password = ${cfg.password}"}
+    ${lib.optionalString (cfg.password_cmd != null) "password_cmd = \"${cfg.password_cmd}\""}
+    ${lib.optionalString cfg.keyring "use_keyring = true"}
+    backend = ${cfg.backend}
+    device = ${cfg.device}
+    ${lib.optionalString (cfg.control_device != null) "control = ${cfg.control_device}"}
+    mixer = ${cfg.mixer}
+    volume_controller = ${cfg.alsa_audio_device}
+    ${lib.optionalString (cfg.song_change_hook != null) "on_song_change_hook = \"${cfg.song_change_hook}\""}
+    device_name = ${cfg.name}
+    bitrate = ${toString cfg.bitrate}
+    cache_path = ${cfg.cache}
+    no_audio_cache = ${toString cfg.audio_cache}
+    volume_normalisation = ${toString cfg.normalize}
+    normalisation_pregain = ${toString cfg.pregain}
+    ${lib.optionalString (cfg.zeroconf != null) "zeroconf_port = ${toString cfg.zeroconf}"}
+  '';
+in with lib; {
+  options.services.spotifyd = {
+    enable = mkEnableOption "SpotifyD connect";
+
+    package = mkOption {
+      type = types.package;
+      default = pkgs.spotifyd;
+      description = "wich spotifyd to use";
+    };
+
+    user = mkOption {
```

Comment:
I think you'd be better off removing all options below and instead having a single `settings` option that you just `toINI` to the configuration file. Sort of like the [programs.noti](https://github.com/rycee/home-manager/blob/master/modules/programs/noti.nix) module but with `toINI` instead of `toYAML`.

Reply:
But then how about the default values? And also with this approach, one can just look up the available options directly, and must not search by spotifyd

Reply:
If you don't specify an option I assume that spotifyd would use its own default value? By specifying many options you become vulnerable to upstream changes. E.g., changes in the upstream default values and changed option names. If the options are not hard coded then the user always is able to cope by changing their configuration.

I don't mind particularly having options but you should be aware that your maintenance burden grows a lot by having them. I would also insist on having an `extraConfig` option to allow the user to cope with such changes.

Oh, the option names should also be changed to camel case as per the contribution guidelines.

Reply:
ok, Used the toINI function